### PR TITLE
docs: init io-descriptors example

### PR DESCRIPTION
**Pull request by lycheel1**
## What does this PR address?

Fixes #(issue)

## Before submitting:

- [x] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
      guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [x] Does the code follow BentoML's code style, `pre-commit run -a` script has passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [x] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [x] Did your changes require updates to the documentation? Have you updated
      those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [x] Did you write tests to cover your changes?

### CODE REVIEW

File: examples/io-descriptors/service.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+from bentoml.validators import DataframeSchema
+from bentoml.validators import DType
+
+# PROMPT_TEMPLATE = """<image>\nUSER: What's the content of the image?\nASSISTANT:"""
+
+
+@bentoml.service()
+class ImageResize:
+    @bentoml.api()
+    def generate(self, image: Image, height: str = "64", width: str = "64") -> Image:
```
Comment (larme) [inline examples/io-descriptors/service.py:21]: why height and width are not type int but str?

File: examples/io-descriptors/service.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+@bentoml.service()
+class AppendStringToFile:
+    @bentoml.api()
+    def append_string_to_eof(
+        self,
+        txt_file: t.Annotated[Path, bentoml.validators.ContentType("text/plain")],
+        input_string: str,
+    ) -> t.Annotated[Path, bentoml.validators.ContentType("text/plain")]:
+        with open(txt_file, "a") as file:
+            file.write(input_string)
```
Comment (larme) [inline examples/io-descriptors/service.py:53]: when writing a file, use a tmp directory like https://github.com/bentoml/BentoXTTS/blob/main/service.py#L37
same for audiospeedup example

File: examples/io-descriptors/README.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+After you finished configuring `bentofile.yaml`, run `bentoml serve .` to deploy the service locally. You can then interact with the auto-generated swagger UI to play
+around with each different API endpoints.
+
+## Different data types
+
+### Standard Python types
+
+The following demonstrates a simple addtion service, with both inputs and output as float parameters. You can
+obviously change the type annotation to `int`, `str` etc. to get familiar with the interaction between type
+annotaions and the auto-generated Swagger UI when deploying locally.\
```
Comment (Sherlock113) [inline examples/io-descriptors/README.md:55]:
```suggestion
The following demonstrates a simple addition Service, with both inputs and output as float parameters. You can
obviously change the type annotation to `int`, `str` etc. to get familiar with the interaction between type
annotations and the auto-generated Swagger UI when deploying locally.
```

File: examples/io-descriptors/README.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+Let's start with the environment. We recommend using virtual environment for better package handling.
+
+```bash
+python -m venv io-descriptors-example
+source io-descriptors-example/bin/activate
+pip install -r requirements.txt
+```
+
+7 different API services are implemented within this example, with diversed input/output types. When running, you should specified the class name of the service
```
Comment (Sherlock113) [inline examples/io-descriptors/README.md:26]:
```suggestion
## Running a Service

7 different API Services are implemented within this example in `service.py`, with diverse input/output types. When running, you should specify the class name of the Service
```

File: examples/io-descriptors/README.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        self,
+        txt_file: t.Annotated[Path, bentoml.validators.ContentType("text/plain")], input_string: str
+        ) -> t.Annotated[Path, bentoml.validators.ContentType("text/plain")]:
+        with open(txt_file, "a") as file:
+            file.write(input_string)
+        return txt_file
+```
+
+Within `service.py`, example API services with 4 different file types are implemented (audio, image, text file, and pdf file). The functionality of each service
+is quite simple and self-explainatory.
```
Comment (Sherlock113) [inline examples/io-descriptors/README.md:88]:
```suggestion
is quite simple and self-explanatory.
```

File: examples/io-descriptors/README.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+`PIL.Image.Image`, which means that image objects can be directly passed through clients, instead of a file object.
+
+The last two services are examples of having `numpy.ndarray` or `pandas.DataFrame` as input parameters. Since they all work quite similarly with the above examples,
+we will not specifically explain them in this tutorial. You can try to write revise the service with `torch.Tensor` as input to check your understanding.
+
+To serve the these examples locally, run `bentoml serve .`
+
+```bash
+bentoml serve .
+2024-03-22T19:25:24+0000 [INFO] [cli] Starting production HTTP BentoServer from "service:ImageResize" listening on http://localhost:3000 (Press CTRL+C to quit)
```
Comment (Sherlock113) [inline examples/io-descriptors/README.md:100]:
```suggestion
$ bentoml serve .

2024-03-22T19:25:24+0000 [INFO] [cli] Starting production HTTP BentoServer from "service:ImageResize" listening on http://localhost:3000 (Press CTRL+C to quit)
```

File: examples/io-descriptors/README.md [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+```
+
+Open your web browser at http://0.0.0.0:3000 to view the Swagger UI for sending test requests.
+
+You may also send request with `curl` command or any HTTP client, e.g.:
+
+```bash
+curl -X 'POST' \
+  'http://0.0.0.0:3000/classify_generated_texts' \
```
Comment (Sherlock113) [inline examples/io-descriptors/README.md:109]: Let's interact with one of the endpoints in this example. `classify_generated_texts` is not included in this example.

File: examples/io-descriptors/README.md [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+- Pydantic field types: see [Pydantic types documentation](https://field-idempotency--pydantic-docs.netlify.app/usage/types/).
+- ML specific types: `nummpy.ndarray`, `torch.Tensor` , `tf.Tensor` for tensor data, `pd.DataFrame` for tabular data, `PIL.Image.Image` for
+Image data, and `pathlib.Path` for files such as audios, images, and pdfs.
+
+When creating a Bentoml service, you should use Python's type annotations to define the expected input and output types for each API endpoint. This
+step can not only helps validate the data against the specified schema, but also enhances the clarity and readability of your code. Type annotations play
+an important role in generating the BentoML API, client, and service UI components, ensuring a consitent and predictable interaction with the Service.
+
+You can also use `pydantic.Field` to set additional information about servicee parameters, such as default values and descriptions. This improves the API's
+usability and provides basic documentation. See the following examples for details.
```
Comment (Sherlock113) [inline examples/io-descriptors/README.md:14]:
```suggestion
usability and provides basic documentation.

In this tutorial, you will learn how to set different [input and output types](https://docs.bentoml.com/en/latest/guides/iotypes.html) for BentoML Services.
```

File: examples/io-descriptors/README.md [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+## Running this example
```
Comment (Sherlock113) [inline examples/io-descriptors/README.md:16]: Use a separate section here.

```suggestion
## Installing dependencies
```

File: examples/io-descriptors/README.md [hunk 10]
```diff
@@ -0,0 +1,120 @@
+# BentoML Input/Output Types Tutorial
+
+A BentoML service supports a wide range of data types when creating a Service API. The data types can be catagorized as follows:
```
Comment (Sherlock113) [inline examples/io-descriptors/README.md:3]:
```suggestion
BentoML supports a wide range of data types when creating a Service API. The data types can be categorized as follows:
```

File: examples/io-descriptors/README.md [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+- Python Standards: `str`, `int`, `float`, `list`, `dict` etc.
+- Pydantic field types: see [Pydantic types documentation](https://field-idempotency--pydantic-docs.netlify.app/usage/types/).
+- ML specific types: `nummpy.ndarray`, `torch.Tensor` , `tf.Tensor` for tensor data, `pd.DataFrame` for tabular data, `PIL.Image.Image` for
+Image data, and `pathlib.Path` for files such as audios, images, and pdfs.
+
+When creating a Bentoml service, you should use Python's type annotations to define the expected input and output types for each API endpoint. This
```
Comment (Sherlock113) [inline examples/io-descriptors/README.md:9]: Change all BentoML `services` to `Services`. "Service" is a special term in BentoML.

```suggestion
When creating a BentoML Service, you should use Python's type annotations to define the expected input and output types for each API endpoint. This
```

**Review by larme — APPROVED** [review]
LGTM, minor changes will be another pr

**Pull request merged**